src/cmd/cmdmap.cpp

```diff
@@ -1,42 +1,131 @@
 #include "cmdmap.h"
 #include "logger.h"
+#include "cmdimp.h"
 
 std::string CommandMap::get_usage(std::string command) {
   // Check if command exists?
+  std::string result;
   try {
-    return info_map[command].usage;
+    result = info_map[command]->usage;
   }
   catch(int e) {
     Logger::info << "get_usage(" << command << ") throws error" << e;
-    exit(1);
+    result = "Error: usage not found";
   }
+  return result;
 }
 
 int CommandMap::exec_command(cmdargs &cmd, std::string command) {
+  // Check if the commands vector contains command
+  bool seen = false;
+  for(unsigned int i = 0; i < commands.size(); i++) {
+    if(command == commands[i])
+      seen = true;
+  }
+  if(!seen) {
+    Logger::fatal << "Unrecognized command: " << command << Logger::endl;
+    return -1;
+  }
+
+  int result;
   try {
-    return info_map[command].helper_fn(cmd);
+    result =  info_map[command]->helper_fn(cmd);
   }
   catch(int e) {
     Logger::info << "exec_command(" << command << ") throws error" << e;
-    exit(1);
+    result = -1;
   }
+  return result;
 }
 
-int CommandMap::add_command(std::string cmd_name, std::string cmd_usage, int (* helper_fn)(cmdargs cmd)) {
-  // Add the command to the vector of commands
-  commands.push_back(cmd_name);
+CommandMap init_autolab_command_map() {
 
-  // Construct the command_info struct for this command
-  command_info cmd_info;
-  cmd_info.usage = cmd_usage;
-  cmd_info.helper_fn = helper_fn;
-  info_map[cmd_name] = cmd_info;
+  std::vector<std::string> valid_commands = {"status", "download", "submit",
+                                             "courses", "assessments", "asmts",
+                                             "problems", "scores", "feedback",
+                                             "enroll"};
```

Comment:
are there any additional uses for the valid_commands vector?
I'm thinking we may be able to omit this, and instead have print_help() loop through the values of the map directly. This will also avoid the need to check if a command exists each time we get its usage string.

Reply:
Yeah, it's not necessary. The main "advantage" of the vector is to make sure we print the values in a specified order, but that's really a matter of preference.

Reply:
I see. I guess it does make sense to be able to order them, but having another vector just seems a bit redundant... What do you think if we replaced the map with an array of command_info structs in order? Would that make it better or worse

Reply:
Huh, that would make it more efficient, but then we don't have the intuition of retrieving command info from the command name. We'd also sacrifice the O(1) lookup time for O(n), but when our n is equal to 10... I guess getting rid of the ordering isn't too big of a loss, if we ditch the redundancy.

Reply:
yeah I wouldn't worry about the performance with 10 elements since the map uses a red-black tree internally (log n), which also needs to be self-balanced during insertion. I do agree that the intuition of "mapping" command names to command functions is quite useful. I guess we can leave the map as is and remove the order vector. Iterating through the map is guaranteed to be in lexicographical order anyway, so it's not bad at all.

Reply:
Oh, that's no problem then. I can take care of fixing this and removing the unnecessary comment below in the morning. I'll commit on top then.

Reply:
sounds good, thanks!